Configure supported cloud backends at compile time

tretter opened the pull request:
## Proposed changes

Use Cargo features to configure the supported cloud backends in the thin-edge mapper and the supported bridges.

An integrator who wants to use thin-edge in a product can use the features to explicitly disable supported cloud backends at compile time.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue

https://github.com/thin-edge/thin-edge.io/issues/3584

<br/>

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s. You can activate automatic signing by running `just prepare-dev` once)
- [ ] I ran `just format` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I used `just check` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

tretter commented:
> For consistency, I'd recommend to also add a feature flag for the Cumulocity (`c8y`) mapper as well, as someone could also have the same requirements as you but want to enforce usage of a different cloud.

I'll add a c8y flag.

> Though I'm assuming that you're also comfortable with building the packaging yourself and don't need a specific thin-edge.io release linux package which only has the desired mappers configured (as this would complicate the release significantly).

I don't need a specific thin-edge.io release package, and I'd also assume that anyone who needs a special configuration of the mappers is also able to build and package thin-edge.io themselves.

albinsuresh commented on crates/core/tedge_mapper/Cargo.toml:
```diff
@@ -38,6 +38,9 @@ tedge_uploader_ext = { workspace = true }
 tracing = { workspace = true }
 
 [features]
+default = ["aws"]
```
I'd either drop that default or define `c8y` as the default, as that's the most commonly used one (provided that `c8y_mapper_ext` is also put behind a feature flag).

albinsuresh replied:
Okay, just saw [this comment](https://github.com/thin-edge/thin-edge.io/pull/3585#issuecomment-2835473791)

tretter commented:
> > For consistency, I'd recommend to also add a feature flag for the Cumulocity (`c8y`) mapper as well, as someone could also have the same requirements as you but want to enforce usage of a different cloud.
> 
> I'll add a c8y flag.

I added a c8y feature to disable c8y, too. This needed a bit more refactoring to cleanly add the necessary attributes.

If **all** features for the supported backend are disabled, the build throws a lot of warnings because of an uninhabited `enum MaybeBorrowedCloud` and errors for all related match statements. Furthermore, disabling all features results in various unused imports. I don't think disabling all cloud support is reasonable. Thus, I added an explicit compiler error to tell the user that at least one of `aws`, `azure` or `c8y` need to be enabled.

didier-wenzek reviewed (approved):
Approved. Thank you for this contribution.

reubenmiller reviewed (approved):
Approved (functionally). Nice to see some new use-cases, thanks for the contribution.
